sql/src/main/java/io/crate/operation/projectors/DMLProjector.java

```diff
@@ -0,0 +1,144 @@
+ * agreement.
+ */
+
+package io.crate.operation.projectors;
+
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import io.crate.core.collections.Row;
+import io.crate.core.collections.Row1;
+import io.crate.executor.transport.ShardRequest;
+import io.crate.executor.transport.TransportActionProvider;
+import io.crate.jobs.ExecutionState;
+import io.crate.operation.collect.CollectExpression;
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.action.bulk.BulkRetryCoordinatorPool;
+import org.elasticsearch.action.bulk.BulkShardProcessor;
+import org.elasticsearch.cluster.ClusterService;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.mapper.Uid;
+import org.elasticsearch.index.shard.ShardId;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.BitSet;
+import java.util.UUID;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class DMLProjector<Request extends ShardRequest> extends AbstractProjector {
+
+    public static final int DEFAULT_BULK_SIZE = 1024;
+
+    private final AtomicReference<Throwable> upstreamFailure = new AtomicReference<>(null);
+
+    private final ShardId shardId;
+    private final CollectExpression<Row, ?> collectUidExpression;
+    private final Object lock = new Object();
+
+    protected final ClusterService clusterService;
+    protected final Settings settings;
+    protected final TransportActionProvider transportActionProvider;
+    protected final BulkRetryCoordinatorPool bulkRetryCoordinatorPool;
+    protected final UUID jobId;
+
+    protected BulkShardProcessor<Request> bulkShardProcessor;
+
+    public DMLProjector(ClusterService clusterService,
+                        Settings settings,
+                        ShardId shardId,
+                        TransportActionProvider transportActionProvider,
+                        BulkRetryCoordinatorPool bulkRetryCoordinatorPool,
+                        CollectExpression<Row, ?> collectUidExpression,
+                        UUID jobId) {
+        this.clusterService = clusterService;
+        this.settings = settings;
+        this.transportActionProvider = transportActionProvider;
+        this.bulkRetryCoordinatorPool = bulkRetryCoordinatorPool;
+        this.shardId = shardId;
+        this.collectUidExpression = collectUidExpression;
+        this.jobId = jobId;
+    }
+
+    protected abstract BulkShardProcessor<Request> createBulkShardProcessor(int bulkSize);
+
+    protected abstract ShardRequest.Item createItem(String id);
+
+    @Override
+    public void prepare(ExecutionState executionState) {
+        super.prepare(executionState);
+        bulkShardProcessor = createBulkShardProcessor(DEFAULT_BULK_SIZE);
+    }
+
+    @Override
+    public boolean setNextRow(Row row) {
+        final Uid uid;
+        synchronized (lock) {
```

Comment:
I think the lock isn't required, each projector either runs in it's own thread or is protected by the RowMerger